### ISSUE TITLE
Link issue with igl::fast_winding_number

### ISSUE THREAD

**Issue by dcoeurjo** [libigl/libigl#895]
First, thanks for sharing your code;)

I'm having issues when building a simple test on fast_winding_number (on macOS). First, I've used @alecjacobson 's trick to downgrade the CGAL version from 4.12 to 4.11 (#794).

The code is the following one:
``` c
#include <iostream>
#include <igl/copyleft/cgal/fast_winding_number.h>
int main(int argc, char *argv[])
{
  const Eigen::MatrixXd points= (Eigen::MatrixXd(3,3)<<
                                 0.0,0.0,0.0,
                                 0.0,0.0,1.0,
                                 0.0,1.0,0.0).finished();
  
  const Eigen::MatrixXd normals= (Eigen::MatrixXd(3,3)<<
                                  0.0,0.0,1.0,
                                  0.0,0.0,1.0,
                                  0.0,0.0,1.0).finished();
  
  const Eigen::MatrixXd probe= (Eigen::MatrixXd(1,3)<<
                                0.5,0.5,0.5).finished();
  
  Eigen::MatrixXd result;
  igl::fast_winding_number(points,normals, probe, 0,2.0, result);
  
  std::cout<< result<<std::endl;
}
```
with the following CMakeLists (https://gist.github.com/dcoeurjo/339bece1597e4a98aab38cf3c315228f) 

When building, I have link issues:

```
[100%] Linking CXX executable example_bin
Undefined symbols for architecture x86_64:
  "void igl::fast_winding_number<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, -1, 0, -1, -1>, double, Eigen::Matrix<double, -1, -1, 0, -1, -1> >(Eigen::MatrixBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> > const&, Eigen::MatrixBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> > const&, Eigen::MatrixBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> > const&, int, double, Eigen::PlainObjectBase<Eigen::Matrix<double, -1, -1, 0, -1, -1> >&)", referenced from:
      _main in main.cpp.o
ld: symbol(s) not found for architecture x86_64
```

It seems to be a `target_link_libraries` issue but the `igl::cgal` should do the job.. 
Any idea ?

(NB: same issue building with static igl)

**Comment by jdumas** [comment]
Hi David,

If I didn't know any better I'd say this looks like a missing template instantiation. Are you sure that `LIBIGL_USE_STATIC_LIBRARY` is `OFF` in you `CMakeCache.txt`?

**Comment by alecjacobson** [comment]
My guess is @jdumas is on the right track about the static library. Either turn it off or [add a template](http://libigl.github.io/libigl/static-library/#explicit-instantiations-of-templated-functions). I already added yours above, so should be merged soon.

Meanwhile, I've (finally) merged the PR https://github.com/libigl/libigl/pull/832 to update CGAL to 4.12. Hopefully this removes the first issue.

**Comment by dcoeurjo** [comment]
Hi Jérémie ! 
I'm pretty sure my build folder was clean but let me try again ;)

@alecjacobson thanks. I'll reset my cgal version

**Comment by dcoeurjo** [comment]
@jdumas : STATIC ON on clean build -> same issue (master before #832 )

@jdumas + @alecjacobson : STATIC OFF (clean build) with the current master -> no more CGAL issue at cmake but same linking error

On the igl master, tutorials (and CGAL tutorials) are still fine.

**Comment by dcoeurjo** [comment]
(I've also used #896 to build the tutorials;))

**Comment by alecjacobson** [comment]
Just to be clear you're passing the flag: `-DLIBIGL_USE_STATIC_LIBRARY=OFF`
and getting this missing symbol error?

**Comment by dcoeurjo** [comment]
Yes (see snap, I have to run the cmake twice due to strange parsing errors during the FindMatlab but it may be not related to this issue).

![tty](https://user-images.githubusercontent.com/700165/45758856-e0bc0e00-bc26-11e8-9697-64a54802c057.gif)

**Comment by alecjacobson** [comment]
Strange... We should also sort out the matlab thing but on a different
issue.

meanwhile, it does seem like there was an incorrect name spacing on the
igl/copyleft/cgal/fast_winding_number.h hopefully this was the issue:

https://github.com/libigl/libigl/commit/373eee9c9c889d92078b1c6fb5c89f703182eacb

**Comment by dcoeurjo** [comment]
mmm.. if I cherry-pick this commit, I get other building errors:

``` 
[ 50%] Building CXX object CMakeFiles/example_bin.dir/main.cpp.o
/Users/davidcoeurjolly/local/src/libigl-example-project/main.cpp:19:3: error: no matching function for call to 'fast_winding_number'
  igl::fast_winding_number(points,normals, probe, 0,2.0, result);
  ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/davidcoeurjolly/local/src/libigl/shared/cmake/../../include/igl/copyleft/cgal/../../fast_winding_number.cpp:12:19: note: candidate function template not viable: requires 9 arguments, but 6 were provided
  IGL_INLINE void fast_winding_number(const Eigen::MatrixBase<DerivedP>& P,
                  ^
/Users/davidcoeurjolly/local/src/libigl/shared/cmake/../../include/igl/copyleft/cgal/../../fast_winding_number.cpp:119:19: note: candidate function template not viable: requires 11 arguments, but 6 were provided
  IGL_INLINE void fast_winding_number(const Eigen::MatrixBase<DerivedP>& P,
                  ^
/Users/davidcoeurjolly/local/src/libigl/shared/cmake/../../include/igl/copyleft/cgal/../../fast_winding_number.cpp:291:19: note: candidate function template not viable: requires 5 arguments, but 6 were provided
  IGL_INLINE void fast_winding_number(const Eigen::MatrixBase<DerivedP>& P,
                  ^
/Users/davidcoeurjolly/local/src/libigl/shared/cmake/../../include/igl/copyleft/cgal/../../fast_winding_number.cpp:264:19: note: candidate function template not viable: requires 7 arguments, but 6 were provided
  IGL_INLINE void fast_winding_number(const Eigen::MatrixBase<DerivedP>& P,
                  ^
1 error generated.
```

**Comment by jdumas** [comment]
Yep, can confirm I get the same behavior on my machine. This new code is not used in any tutorial yet so the Travis CI build wouldn't catch this kind of issue (esp. when missing template instantiations).

With Alec's new header you just need to modify your call to use the function in the `igl::copyleft::cgal` namespace`:

```cpp
igl::copyleft::cgal::fast_winding_number(points,normals, probe, 0,2.0, result);
```

**Comment by dcoeurjo** [comment]
excellent ! thank you, guys.

I'll open another issue for the matlab stuff.. thanks again @jdumas & @alecjacobson

**Comment by dcoeurjo** [comment]
Just quick question: is the commit you mentioned  in a pending PR ?
